### ISSUE TITLE
virtio-gpu device is not supported by qemu-kvm on Arm64

### ISSUE THREAD

**Issue by zhlhahaha** [kubevirt/kubevirt#8708]
**What happened**:
Recently, I do e2e test on Arm64 server, and meet many failures caused by virtio-gpu.
The root cause is qemu-kvm binary not support virtio-gpu any more. And this binary used to support virtio-gpu. Are there any reasons for remove this device?

**What you expected to happen**:
qemu-kvm binary support virtio-gpu

**How to reproduce it (as minimally and precisely as possible)**:
```
 docker run --rm -it --entrypoint=/bin/bash quay.io/kubevirt/virt-launcher:20221101_03e057197-arm64
/usr/libexec/qemu-kvm -M virt -device help
...
Display devices:
name "ramfb", bus System, desc "ram framebuffer standalone device"
```

cc: @rmohr @andreabolognani @alicefr

**Comment by kvaps** [comment]
If I understand it correct this can be enabled by specific option when compiling.
Unfortunately we do not maintain qemu package, try to file a bug report to https://bugzilla.redhat.com/

**Comment by andreabolognani** [comment]
Never mind, it's not been disabled: it's just been moved to its own separate package (`qemu-kvm-device-display-virtio-gpu-pci`) as part of the effort to make QEMU more modular.

I will make it so the new package is included in the virt-launcher image, but first let me try to figure out whether there are any other devices that are used by KubeVirt and that have been similarly moved to separate packages in CentOS Stream 9.

@zhlhahaha is there a specific upstream functional test that has allowed to notice this? Does it only run on aarch64? Can we perhaps introduce a similar test that runs on x86_64, to make sure such a regression can't sneak in again?

**Comment by andreabolognani** [comment]
> Can we perhaps introduce a similar test that runs on x86_64, to make sure such a regression can't sneak in again?

This is going to be challenging, as it appears that currently KubeVirt doesn't really offer any way to configure the video device(s) explicitly: you either get the default one (`model=vga` for x86_64, `model=virtio` for aarch64) or [disable graphics altogether](https://kubevirt.io/user-guide/virtual_machines/virtual_hardware/#video-and-graphics-device) and get nothing.

A configuration such as

```yaml
spec:
  domain:
    devices:
      autoattachGraphicsDevice: false
      video:
      - video:
          model:
            type: virtio
```

looks like it would work and result in `model=virtio` being used even for x86_64, but instead it gets rejected with

```
error validating data: ValidationError(VirtualMachineInstance.spec.domain.devices): unknown field "video" in io.kubevirt.v1.VirtualMachineInstance.spec.domain.devices; if you choose to ignore these errors, turn validation off with --validate=false
```

Turning off validation as suggested results in the VM being created but having no video device.

I wonder if this mechanism should be made more flexible, allowing the YAML snipped above to work as expected.

**Comment by zhlhahaha** [comment]
> I wonder if this mechanism should be made more flexible, allowing the YAML snipped above to work as expected.

The vga devices is not support by qemu-kvm binary on Arm64, so we set virtio-gpu as default graphic device. 
I do not have much experience on running VM workload in production. Are there many people use virtio-gpu?

**Comment by andreabolognani** [comment]
> The vga devices is not support by qemu-kvm binary on Arm64, so we set virtio-gpu as default graphic device.

Yeah, that's fine. It's the only reasonable default for the architecture.

> I do not have much experience on running VM workload in production. Are there many people use virtio-gpu?

I guess mostly those who run aarch64 VMs. libvirt only uses virtio-gpu on x86_64 for non-primary displays, so chances of encountering it in the wild are probably quite low.

**Comment by zhlhahaha** [comment]
There are still some problems on booting VM on Arm64 after update Centos Stream 9. I run a simple test `test_id:1620` on Arm64 server.
I got following messages from virt-launcher. If I checkout to the commit `66157f8afa1d03dc06080d721085d743229e89e6`, one commit before `Switch to CentOS Stream 9`, and re-create the kubevirt cluster, the test has passed. 
```
{"component":"virt-launcher","kind":"","level":"info","msg":"Domain XML generated. Base64 dump PGRvbWFpbiB0eXBlPSJrdm0iIHhtbG5zOnFlbXU9Imh0dHA6Ly9saWJ2aXJ0Lm9yZy9zY2hlbWFzL2RvbWFpbi9xZW11LzEuMCI+Cgk8bmFtZT5rdWJldmlydC10ZXN0LWRlZmF1bHQxX3Rlc3R2bWktdHhoenQ8L25hbWU+Cgk8bWVtb3J5IHVuaXQ9ImIiPjI2ODQzNTQ1NjwvbWVtb3J5PgoJPG9zPgoJCTx0eXBlIGFyY2g9ImFhcmNoNjQiIG1hY2hpbmU9InZpcnQiPmh2bTwvdHlwZT4KCQk8bG9hZGVyIHJlYWRvbmx5PSJ5ZXMiIHNlY3VyZT0ibm8iIHR5cGU9InBmbGFzaCI+L3Vzci9zaGFyZS9BQVZNRi9BQVZNRl9DT0RFLmZkPC9sb2FkZXI+CgkJPG52cmFtIHRlbXBsYXRlPSIvdXNyL3NoYXJlL0FBVk1GL0FBVk1GX1ZBUlMuZmQiPi90bXAva3ViZXZpcnQtdGVzdC1kZWZhdWx0MV90ZXN0dm1pLXR4aHp0PC9udnJhbT4KCTwvb3M+Cgk8c3lzaW5mbyB0eXBlPSJzbWJpb3MiPgoJCTxzeXN0ZW0+CgkJCTxlbnRyeSBuYW1lPSJ1dWlkIj45MGU4MzJiNS0yN2VkLTQ1NjMtYTQ5NS00ZmFiZjgxYTZhNGU8L2VudHJ5PgoJCQk8ZW50cnkgbmFtZT0ibWFudWZhY3R1cmVyIj5LdWJlVmlydDwvZW50cnk+CgkJCTxlbnRyeSBuYW1lPSJmYW1pbHkiPkt1YmVWaXJ0PC9lbnRyeT4KCQkJPGVudHJ5IG5hbWU9InByb2R1Y3QiPk5vbmU8L2VudHJ5PgoJCQk8ZW50cnkgbmFtZT0ic2t1Ij48L2VudHJ5PgoJCQk8ZW50cnkgbmFtZT0idmVyc2lvbiI+PC9lbnRyeT4KCQk8L3N5c3RlbT4KCQk8Ymlvcz48L2Jpb3M+CgkJPGJhc2VCb2FyZD48L2Jhc2VCb2FyZD4KCQk8Y2hhc3Npcz48L2NoYXNzaXM+Cgk8L3N5c2luZm8+Cgk8ZGV2aWNlcz4KCQk8aW50ZXJmYWNlIHR5cGU9ImV0aGVybmV0Ij4KCQkJPHNvdXJjZT48L3NvdXJjZT4KCQkJPHRhcmdldCBkZXY9InRhcDAiIG1hbmFnZWQ9Im5vIj48L3RhcmdldD4KCQkJPG1vZGVsIHR5cGU9InZpcnRpby1ub24tdHJhbnNpdGlvbmFsIj48L21vZGVsPgoJCQk8bXR1IHNpemU9IjE1MDAiPjwvbXR1PgoJCQk8YWxpYXMgbmFtZT0idWEtZGVmYXVsdCI+PC9hbGlhcz4KCQkJPHJvbSBlbmFibGVkPSJubyI+PC9yb20+CgkJPC9pbnRlcmZhY2U+CgkJPGNoYW5uZWwgdHlwZT0idW5peCI+CgkJCTx0YXJnZXQgbmFtZT0ib3JnLnFlbXUuZ3Vlc3RfYWdlbnQuMCIgdHlwZT0idmlydGlvIj48L3RhcmdldD4KCQk8L2NoYW5uZWw+CgkJPGNvbnRyb2xsZXIgdHlwZT0idXNiIiBpbmRleD0iMCIgbW9kZWw9InFlbXUteGhjaSI+PC9jb250cm9sbGVyPgoJCTxjb250cm9sbGVyIHR5cGU9InNjc2kiIGluZGV4PSIwIiBtb2RlbD0idmlydGlvLW5vbi10cmFuc2l0aW9uYWwiPjwvY29udHJvbGxlcj4KCQk8Y29udHJvbGxlciB0eXBlPSJ2aXJ0aW8tc2VyaWFsIiBpbmRleD0iMCIgbW9kZWw9InZpcnRpby1ub24tdHJhbnNpdGlvbmFsIj48L2NvbnRyb2xsZXI+CgkJPHZpZGVvPgoJCQk8bW9kZWwgdHlwZT0idmlydGlvIiBoZWFkcz0iMSI+PC9tb2RlbD4KCQk8L3ZpZGVvPgoJCTxncmFwaGljcyB0eXBlPSJ2bmMiPgoJCQk8bGlzdGVuIHR5cGU9InNvY2tldCIgc29ja2V0PSIvdmFyL3J1bi9rdWJldmlydC1wcml2YXRlLzFlYzhhZTc0LWQ2NGQtNGMxYy1hYjA2LTE0ZDZhZTVmZDVhMi92aXJ0LXZuYyI+PC9saXN0ZW4+CgkJPC9ncmFwaGljcz4KCQk8bWVtYmFsbG9vbiBtb2RlbD0idmlydGlvLW5vbi10cmFuc2l0aW9uYWwiPgoJCQk8c3RhdHMgcGVyaW9kPSIxMCI+PC9zdGF0cz4KCQk8L21lbWJhbGxvb24+CgkJPGRpc2sgZGV2aWNlPSJkaXNrIiB0eXBlPSJmaWxlIiBtb2RlbD0idmlydGlvLW5vbi10cmFuc2l0aW9uYWwiPgoJCQk8c291cmNlIGZpbGU9Ii92YXIvcnVuL2t1YmV2aXJ0LWVwaGVtZXJhbC1kaXNrcy9kaXNrLWRhdGEvZGlzazAvZGlzay5xY293MiI+PC9zb3VyY2U+CgkJCTx0YXJnZXQgYnVzPSJ2aXJ0aW8iIGRldj0idmRhIj48L3RhcmdldD4KCQkJPGRyaXZlciBjYWNoZT0ibm9uZSIgZXJyb3JfcG9saWN5PSJzdG9wIiBuYW1lPSJxZW11IiB0eXBlPSJxY293MiIgZGlzY2FyZD0idW5tYXAiPjwvZHJpdmVyPgoJCQk8YWxpYXMgbmFtZT0idWEtZGlzazAiPjwvYWxpYXM+CgkJCTxiYWNraW5nU3RvcmUgdHlwZT0iZmlsZSI+CgkJCQk8Zm9ybWF0IHR5cGU9InJhdyI+PC9mb3JtYXQ+CgkJCQk8c291cmNlIGZpbGU9Ii92YXIvcnVuL2t1YmV2aXJ0L2NvbnRhaW5lci1kaXNrcy9kaXNrXzAuaW1nIj48L3NvdXJjZT4KCQkJPC9iYWNraW5nU3RvcmU+CgkJCTxleHBhbmREaXNrc0VuYWJsZWQ+dHJ1ZTwvZXhwYW5kRGlza3NFbmFibGVkPgoJCTwvZGlzaz4KCQk8aW5wdXQgdHlwZT0idGFibGV0IiBidXM9InVzYiI+PC9pbnB1dD4KCQk8aW5wdXQgdHlwZT0ia2V5Ym9hcmQiIGJ1cz0idXNiIj48L2lucHV0PgoJCTxzZXJpYWwgdHlwZT0idW5peCI+CgkJCTx0YXJnZXQgcG9ydD0iMCI+PC90YXJnZXQ+CgkJCTxzb3VyY2UgbW9kZT0iYmluZCIgcGF0aD0iL3Zhci9ydW4va3ViZXZpcnQtcHJpdmF0ZS8xZWM4YWU3NC1kNjRkLTRjMWMtYWIwNi0xNGQ2YWU1ZmQ1YTIvdmlydC1zZXJpYWwwIj48L3NvdXJjZT4KCQk8L3NlcmlhbD4KCQk8Y29uc29sZSB0eXBlPSJwdHkiPgoJCQk8dGFyZ2V0IHR5cGU9InNlcmlhbCIgcG9ydD0iMCI+PC90YXJnZXQ+CgkJPC9jb25zb2xlPgoJPC9kZXZpY2VzPgoJPG1ldGFkYXRhPgoJCTxrdWJldmlydCB4bWxucz0iaHR0cDovL2t1YmV2aXJ0LmlvIj4KCQkJPHVpZD4xZWM4YWU3NC1kNjRkLTRjMWMtYWIwNi0xNGQ2YWU1ZmQ1YTI8L3VpZD4KCQkJPGdyYWNlcGVyaW9kPgoJCQkJPGRlbGV0aW9uR3JhY2VQZXJpb2RTZWNvbmRzPjA8L2RlbGV0aW9uR3JhY2VQZXJpb2RTZWNvbmRzPgoJCQk8L2dyYWNlcGVyaW9kPgoJCTwva3ViZXZpcnQ+Cgk8L21ldGFkYXRhPgoJPGZlYXR1cmVzPgoJCTxhY3BpPjwvYWNwaT4KCTwvZmVhdHVyZXM+Cgk8Y3B1IG1vZGU9Imhvc3QtcGFzc3Rocm91Z2giPgoJCTx0b3BvbG9neSBzb2NrZXRzPSIxIiBjb3Jlcz0iMSIgdGhyZWFkcz0iMSI+PC90b3BvbG9neT4KCTwvY3B1PgoJPHZjcHUgcGxhY2VtZW50PSJzdGF0aWMiPjE8L3ZjcHU+Cgk8aW90aHJlYWRzPjE8L2lvdGhyZWFkcz4KPC9kb21haW4+","name":"testvmi-txhzt","namespace":"kubevirt-test-default1","pos":"libvirt_helper.go:134","timestamp":"2022-11-03T06:04:40.668408Z","uid":"1ec8ae74-d64d-4c1c-ab06-14d6ae5fd5a2"}
{"component":"virt-launcher-monitor","level":"info","msg":"Reaped pid 68 with status 0","pos":"virt-launcher-monitor.go:124","timestamp":"2022-11-03T06:04:40.742146Z"}
{"component":"virt-launcher","level":"error","msg":"Unable to read from monitor: Connection reset by peer","pos":"qemuMonitorIORead:423","subcomponent":"libvirt","thread":"65","timestamp":"2022-11-03T06:04:41.208000Z"}
{"component":"virt-launcher","level":"warning","msg":"Failed to probe capabilities for /usr/libexec/qemu-kvm: Unable to read from monitor: Connection reset by peer","pos":"virQEMUCapsLogProbeFailure:5462","subcomponent":"libvirt","thread":"39","timestamp":"2022-11-03T06:04:41.208000Z"}
{"component":"virt-launcher-monitor","level":"info","msg":"Reaped pid 70 with status 134","pos":"virt-launcher-monitor.go:124","timestamp":"2022-11-03T06:04:41.210177Z"}
{"code":"virErrCode(1)","component":"virt-launcher","level":"error","msg":"Connection to libvirt lost.","pos":"libvirt.go:453","reason":"virError(Code=1, Domain=0, Message='An error occurred, but the cause is unknown')","timestamp":"2022-11-03T06:04:41.410812Z"}
{"component":"virt-launcher","kind":"","level":"error","msg":"Defining the VirtualMachineInstance failed.","name":"testvmi-txhzt","namespace":"kubevirt-test-default1","pos":"libvirt_helper.go:137","reason":"virError(Code=1, Domain=0, Message='An error occurred, but the cause is unknown')","timestamp":"2022-11-03T06:04:41.410922Z","uid":"1ec8ae74-d64d-4c1c-ab06-14d6ae5fd5a2"}
{"component":"virt-launcher","kind":"","level":"error","msg":"Failed to sync vmi","name":"testvmi-txhzt","namespace":"kubevirt-test-default1","pos":"server.go:185","reason":"virError(Code=1, Domain=0, Message='An error occurred, but the cause is unknown')","timestamp":"2022-11-03T06:04:41.410996Z","uid":"1ec8ae74-d64d-4c1c-ab06-14d6ae5fd5a2"}
```

There is no useful information from libvirt log. Maybe caused by crash of qemu-kvm. 
I have check the supported devices of qemu-kvm in Centos Stream9, all necessary devices have been supported.
I also compared the Virtual Machine XML configuration before and after switching to Centos Stream9, they are same.
I am still looking into it. Do you have any suggestion?

**Comment by andreabolognani** [comment]
> There are still some problems on booting VM on Arm64 after update Centos Stream 9. I run a simple test `test_id:1620` on Arm64 server.

That looks like a very basic test, so the fact that it fails is quite concerning.

Can you successfully start any VM at all? Or even outside of the test suite, starting VMs simply doesn't work?

> If I checkout to the commit `66157f8afa1d03dc06080d721085d743229e89e6`, one commit before `Switch to CentOS Stream 9`, and re-create the kubevirt cluster, the test has passed.

Good sanity check.

> There is no useful information from libvirt log.

Have you tried setting `debugLog` and possibly also `logVerbosity`? That might produce more information.

> ```
> {"component":"virt-launcher","level":"warning","msg":"Failed to probe capabilities for /usr/libexec/qemu-kvm: Unable to read from monitor: Connection reset by peer","pos":"virQEMUCapsLogProbeFailure:5462","subcomponent":"libvirt","thread":"39","timestamp":"2022-11-03T06:04:41.208000Z"}
> ```
> Maybe caused by crash of qemu-kvm.

Yeah, the error message seems to indicate that the QEMU process started by libvirt to figure out emulator capabilities didn't run successfully. With a more verbose log, you should be able to fish out the exact command line and QMP interactions. You could then try running the same steps outside of libvirt to see whether the failure still occurs, and generally to get closer to the root cause.

Thanks a lot for looking into this @zhlhahaha!

**Comment by awels** [comment]
IIRC at the top of the virt-launcher log there is a base64 encoded version of the domain xml.

**Comment by andreabolognani** [comment]
> IIRC at the top of the virt-launcher log there is a base64 encoded version of the domain xml.

Yeah, in this case that'd be

```xml
<domain xmlns:qemu="http://libvirt.org/schemas/domain/qemu/1.0" type="kvm">
  <name>kubevirt-test-default1_testvmi-txhzt</name>
  <memory unit="b">268435456</memory>
  <os>
    <type arch="aarch64" machine="virt">hvm</type>
    <loader readonly="yes" secure="no" type="pflash">/usr/share/AAVMF/AAVMF_CODE.fd</loader>
    <nvram template="/usr/share/AAVMF/AAVMF_VARS.fd">/tmp/kubevirt-test-default1_testvmi-txhzt</nvram>
  </os>
  <sysinfo type="smbios">
    <system>
      <entry name="uuid">90e832b5-27ed-4563-a495-4fabf81a6a4e</entry>
      <entry name="manufacturer">KubeVirt</entry>
      <entry name="family">KubeVirt</entry>
      <entry name="product">None</entry>
      <entry name="sku"/>
      <entry name="version"/>
    </system>
    <bios/>
    <baseBoard/>
    <chassis/>
  </sysinfo>
  <devices>
    <interface type="ethernet">
      <source/>
      <target dev="tap0" managed="no"/>
      <model type="virtio-non-transitional"/>
      <mtu size="1500"/>
      <alias name="ua-default"/>
      <rom enabled="no"/>
    </interface>
    <channel type="unix">
      <target name="org.qemu.guest_agent.0" type="virtio"/>
    </channel>
    <controller type="usb" index="0" model="qemu-xhci"/>
    <controller type="scsi" index="0" model="virtio-non-transitional"/>
    <controller type="virtio-serial" index="0" model="virtio-non-transitional"/>
    <video>
      <model type="virtio" heads="1"/>
    </video>
    <graphics type="vnc">
      <listen type="socket" socket="/var/run/kubevirt-private/1ec8ae74-d64d-4c1c-ab06-14d6ae5fd5a2/virt-vnc"/>
    </graphics>
    <memballoon model="virtio-non-transitional">
      <stats period="10"/>
    </memballoon>
    <disk device="disk" type="file" model="virtio-non-transitional">
      <source file="/var/run/kubevirt-ephemeral-disks/disk-data/disk0/disk.qcow2"/>
      <target bus="virtio" dev="vda"/>
      <driver cache="none" error_policy="stop" name="qemu" type="qcow2" discard="unmap"/>
      <alias name="ua-disk0"/>
      <backingStore type="file">
        <format type="raw"/>
        <source file="/var/run/kubevirt/container-disks/disk_0.img"/>
      </backingStore>
      <expandDisksEnabled>true</expandDisksEnabled>
    </disk>
    <input type="tablet" bus="usb"/>
    <input type="keyboard" bus="usb"/>
    <serial type="unix">
      <target port="0"/>
      <source mode="bind" path="/var/run/kubevirt-private/1ec8ae74-d64d-4c1c-ab06-14d6ae5fd5a2/virt-serial0"/>
    </serial>
    <console type="pty">
      <target type="serial" port="0"/>
    </console>
  </devices>
  <metadata>
    <kubevirt xmlns="http://kubevirt.io">
      <uid>1ec8ae74-d64d-4c1c-ab06-14d6ae5fd5a2</uid>
      <graceperiod>
        <deletionGracePeriodSeconds>0</deletionGracePeriodSeconds>
      </graceperiod>
    </kubevirt>
  </metadata>
  <features>
    <acpi/>
  </features>
  <cpu mode="host-passthrough">
    <topology sockets="1" cores="1" threads="1"/>
  </cpu>
  <vcpu placement="static">1</vcpu>
  <iothreads>1</iothreads>
</domain>
```

but @zhlhahaha already confirmed that the generated XML didn't change with the switch to CentOS Stream 9. This is consistent with my expectations.

The error appears to occur while probing the QEMU binary to figure out its capabilities, which is something that isn't affected by the exact contents of the XML anyway.

**Comment by zhlhahaha** [comment]
> @zhlhahaha any luck with your investigation? I'd like to fix this before the next release is tagged, to avoid regressions on aarch64.

Not yet, I will have a presentation this Thursday, and I am working on it. After this, I will continue to fixing this issue. When will next release publish?

**Comment by andreabolognani** [comment]
> When will next release publish?

@zhlhahaha according to https://github.com/kubernetes/sig-release/tree/master/releases/release-1.26 the release date for Kubernetes 1.26 is going to be 2022-12-06, and according to https://groups.google.com/g/kubevirt-dev/c/1l58t0e7fww KubeVirt 0.59.0 should be out shortly (1-4 weeks) after that.

So there's some time... But not _a lot_ of time ;)

**Comment by andreabolognani** [comment]
> Do we have code freeze or test freeze like Kubernetes release procedure?

No idea, sorry O:-)

**Comment by xpivarc** [comment]
@zhlhahaha You should have at least 7 more weeks. More info https://github.com/kubevirt/kubevirt/blob/main/docs/release.md

**Comment by zhlhahaha** [comment]
> @zhlhahaha You should have at least 7 more weeks. More info https://github.com/kubevirt/kubevirt/blob/main/docs/release.md

Thanks, great. I will try to figure out it as soon as possible.

**Comment by zhlhahaha** [comment]
Some update:

1. I modified the output log level of libvirt (make it can print QEMU), but does not get much useful information.
2. I compiled version 7.1 static qemu-kvm binary from source code and replace it with the qemu-kvm in virt-launcher. VM can successfully boot. 

I am checking the rpm source in http://mirror.stream.centos.org/9-stream/AppStream/source/tree/Packages/ to see how qemu-kvm-core rpm be compiled, and try to repeat the compile process to see if I can reproduce the issue.
